File: 606-construct-string-from-binary-tree/606-construct-string-from-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def tree2str(self, root: Optional[TreeNode]) -> str:
        node_parts = [f"{root.val}"]
        
        if root.left is not None:
            node_parts.append(f"({self.tree2str(root.left)})")
            
        if root.right is not None:
            if root.left is None:
                node_parts.append("()")
            node_parts.append(f"({self.tree2str(root.right)})")
            
        return ''.join(node_parts)

File: 606-construct-string-from-binary-tree/606-construct-string-from-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string}
 */
var tree2str = function(root) {
    if (!root) {
        return ''
    }
    
    const repr = '' + root.val
    
    const left = root.left || root.right
        ? `(${tree2str(root.left)})`
        : ''
    const right = root.right
        ? `(${tree2str(root.right)})`
        : ''
    
    return repr + left + right
};